ls1intum/Artemis
# `Tutorial groups`: Allow students to view tutorial groups and tutors to register students

### Checklist
#### General


#### Server
#### Client

### Motivation and Context



The following user stories are implemented:
- As a student I want to check to which tutorial groups I am registered, so I know when and where I have to attend.
- As a student I want to see all tutorial groups, so I possibly can contact another tutor to change my tutorial group.
- As a tutor of a tutorial group, I want to register students to my group, so I do not have to contact the instructor for this every time.
- As a tutor of a tutorial group, I want to derigster students from my group, so I do not have to contact the instructor for this every time.


### Description


_Changes to Client Code_:
- Implemented the course tutorial group overview and detail views with as little code duplication as possible by making previous components generic and reusable.
- Allowed tutors to also access tutorial group management with managing registrations for their own groups as the only functionality.


_Changes to Server Code_:
- Made the `getOne` and `getAllOfCourse` endpoints available for students and tutors. Privacy sensitive information like the actual registrations of students is removed before sending the information to the client if the user is not at least the tutor of the tutorial group.
- Made the `register` and `deregister` endpoints available for the tutor of a tutorial group.
- Added transient properties to the `TutorialGroup` entity, to tell the client if a user is registered or the assigned tutor to a respective tutorial group and the total number of registered students.

_Changes Related to Client Testing:_
- Vastly improved client test setup and coverage


### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with Complaints enabled

1. Log in to Artemis as an Administrator
2. Activate the Tutorial Group Feature Toggle in the admin settings
3. Create Multiple Tutorial Groups and register some students for them
4. Login as one of the registered students and confirm that you see your registered tutorial group and all othe tutorial groups (see screenshots)

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage



Basically all test changes on the client

| Class/File | Branch | Line |
|------------|-------:|-----:|
| course-tutorial-groups.component.ts | 81% | 89% |
| course-tutorial-group-card.component.ts | 90% | 100% |
| course-tutorial-groups-overview.component.ts | 75% | 100% |
| course-tutorial-groups-registered.component.ts| 100% | 100% |
| tutorial-group-detail.component.ts | 100% | 100% |
| tutorial-groups-table.component.ts | 100% | 100% |

### Screenshots



**Student View**

![image](https://user-images.githubusercontent.com/29718932/194070438-248096c2-19f9-4b57-97ae-9a12d802b283.png)
![image](https://user-images.githubusercontent.com/29718932/194070514-6cd0bb0a-da83-467d-a796-6c451a36bcf6.png)
![image](https://user-images.githubusercontent.com/29718932/194070563-6980a8a2-fcbc-43a3-9ed0-e900a3970599.png)
![image](https://user-images.githubusercontent.com/29718932/194070623-996bc3a5-5695-478d-b80b-9ddaa2dd125c.png)
![image](https://user-images.githubusercontent.com/29718932/194070760-4b8841c0-7339-45cc-ad51-7e1edb1e87b8.png)
![image](https://user-images.githubusercontent.com/29718932/194070817-d37f96f5-ff70-4713-b1b4-eff31ca2ca4f.png)

**Tutor View**
![image](https://user-images.githubusercontent.com/29718932/195456486-679ca848-078a-455b-84be-c465b082024e.png)
![image](https://user-images.githubusercontent.com/29718932/195456514-8613f75c-2850-4497-933c-46b9dba8ab23.png)
![image](https://user-images.githubusercontent.com/29718932/195456537-3d574035-5551-4765-8b8b-b0cde8ac815f.png)
